tests/appsec/test_automated_user_and_session_tracking.py

```diff
@@ -0,0 +1,187 @@
+        )
+
+    def test_user_blocking_sdk(self):
+        assert self.config_state_1[rc.RC_STATE] == rc.ApplyState.ACKNOWLEDGED
+        assert self.config_state_2[rc.RC_STATE] == rc.ApplyState.ACKNOWLEDGED
+        interfaces.library.assert_waf_attack(self.r_login_blocked, rule="block-users")
+        assert self.r_login_blocked.status_code == 403
+
+
+@rfc("https://docs.google.com/document/d/1RT38U6dTTcB-8muiYV4-aVDCsT_XrliyakjtAPyjUpw")
+@features.user_monitoring
+class Test_Automated_Session_Tracking:
+    def setup_session_tracking(self):
+        self.r_create_session = weblog.get("/session/new")
+        self.r_home = weblog.get("/", cookies=self.r_create_session.cookies,)
+
+    def test_session_tracking(self):
+        assert self.r_create_session.status_code == 200
+        expected_session = self.r_create_session.text
+
+        assert self.r_home.status_code == 200
+        for _, trace, span in interfaces.library.get_spans(request=self.r_home):
+            meta = span.get("meta", {})
+            assert meta["usr.session_id"] == expected_session
+
+
+BLOCK_SESSION = (
+    "datadog/2/ASM_DD/rules/config",
+    {
+        "version": "2.1",
+        "metadata": {"rules_version": "1.2.6"},
+        "rules": [
+            {
+                "id": "block-sessions",
+                "name": "Block Session Addresses",
+                "tags": {"type": "block_user", "category": "security_response"},
+                "conditions": [
+                    {
+                        "parameters": {"inputs": [{"address": "usr.session_id"}], "data": "blocked_sessions"},
+                        "operator": "exact_match",
+                    }
+                ],
+                "transformers": [],
+                "on_match": ["block"],
+            }
+        ],
+        "rules_data": [{"id": "blocked_sessions", "type": "data_with_expiration", "data": []},],
+    },
+)
+
+
+@rfc("https://docs.google.com/document/d/1RT38U6dTTcB-8muiYV4-aVDCsT_XrliyakjtAPyjUpw")
+@features.user_monitoring
+@scenarios.appsec_runtime_activation
+class Test_Automated_Session_Blocking:
+    def setup_session_blocking(self):
+        rc.rc_state.reset().apply()
+
+        self.config_state_1 = rc.rc_state.set_config(*CONFIG_ENABLED).apply()
+        self.r_create_session = weblog.get("/session/new")
+
+        BLOCK_SESSION[1]["rules_data"][0]["data"].append({"value": self.r_create_session.text, "expiration": 0})
+        self.config_state_2 = rc.rc_state.set_config(*BLOCK_SESSION).apply()
+        self.r_home_blocked = weblog.get("/", cookies=self.r_create_session.cookies,)
+
+    def test_session_blocking(self):
+        assert self.config_state_1[rc.RC_STATE] == rc.ApplyState.ACKNOWLEDGED
+        assert self.r_create_session.status_code == 200
+
+        assert self.config_state_2[rc.RC_STATE] == rc.ApplyState.ACKNOWLEDGED
+        interfaces.library.assert_waf_attack(self.r_home_blocked, rule="block-sessions")
```

Comment:
should we also test the value to "self.r_create_session.text" although it might have been obfuscated by the waf?

Reply:
Yep, done, included in the pattern, thanks for the input